test/csuite/wt7989_compact_checkpoint/main.c

```diff
@@ -0,0 +1,286 @@
+};
+
+/* Forward declarations. */
+static void run_test(const char *home, const char *uri, bool stress_test);
+static void *thread_func_compact(void *arg);
+static void *thread_func_checkpoint(void *arg);
+static void populate(WT_SESSION *session, const char *uri);
+static void remove_records(WT_SESSION *session, const char *uri);
+static uint64_t get_file_size(WT_SESSION *session, const char *uri);
+static void set_timing_stress_checkpoint(WT_CONNECTION *conn);
+
+/* Methods implementation. */
+int
+main(int argc, char *argv[])
+{
+    TEST_OPTS *opts, _opts;
+    char home_cv[512];
+
+    opts = &_opts;
+    memset(opts, 0, sizeof(*opts));
+    testutil_check(testutil_parse_opts(argc, argv, opts));
+
+    /*
+     * First run test with WT_TIMING_STRESS_CHECKPOINT_SLOW.
+     */
+    run_test(opts->home, opts->uri, true);
+
+    /*
+     * Now run test where compact and checkpoint threads are synchronized using condition variable.
+     */
+    testutil_assert(sizeof(home_cv) > strlen(opts->home) + 3);
+    sprintf(home_cv, "%s.CV", opts->home);
+    run_test(home_cv, opts->uri, false);
+
+    /* Cleanup */
+    if (!opts->preserve)
+        testutil_clean_work_dir(home_cv);
+
+    testutil_cleanup(opts);
+
+    return (EXIT_SUCCESS);
+}
+
+static void
+run_test(const char *home, const char *uri, bool stress_test)
+{
+    struct thread_data td;
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    pthread_t thread_checkpoint, thread_compact;
+    uint64_t file_sz_after, file_sz_before;
+
+    testutil_make_work_dir(home);
+    testutil_check(wiredtiger_open(home, NULL, conn_config, &conn));
+
+    if (stress_test) {
+        /*
+         * Set WT_TIMING_STRESS_CHECKPOINT_SLOW flag. It adds 10 seconds sleep before each
+         * checkpoint.
+         */
+        set_timing_stress_checkpoint(conn);
+    }
+
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create and populate dataset. Checkpoint the data after that. */
+    testutil_check(session->create(session, uri, table_config));
+
+    populate(session, uri);
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range. To let compact relocate blocks from the
+     * end of the file.
+     */
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    /* Create and initialize conditional variable. */
+    td.conn = conn;
+    td.uri = uri;
+    if (stress_test) {
+        td.cond = NULL;
+        td.session_cond = NULL;
+    } else {
+        td.session_cond = (WT_SESSION_IMPL *)session;
+        testutil_check(__wt_cond_alloc(td.session_cond, "compact operation", &td.cond));
+    }
+
+    /* Spawn checkpoint and compact threads. */
+    testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+    testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+    (void)pthread_join(thread_checkpoint, NULL);
```

Comment:
add comment : Wait for the threads to finish work.

Reply:
Done